application/src/main/java/org/togetherjava/tjbot/commands/help/AutoPruneHelperRoutine.java

```diff
@@ -0,0 +1,168 @@
+
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.entities.Role;
+import net.dv8tion.jda.api.entities.TextChannel;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.commands.Routine;
+import org.togetherjava.tjbot.config.Config;
+import org.togetherjava.tjbot.db.Database;
+import org.togetherjava.tjbot.moderation.ModAuditLogWriter;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.Period;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static org.togetherjava.tjbot.db.generated.tables.HelpChannelMessages.HELP_CHANNEL_MESSAGES;
+
+/**
+ * Due to a technical limitation in Discord, roles with more than 100 users can not be ghost-pinged
+ * into helper threads.
+ * <p>
+ * This routine mitigates the problem by automatically pruning inactive users from helper roles
+ * approaching this limit.
+ */
+public final class AutoPruneHelperRoutine implements Routine {
+    private static final Logger logger = LoggerFactory.getLogger(AutoPruneHelperRoutine.class);
+
+    private static final int ROLE_FULL_LIMIT = 100;
+    private static final int ROLE_FULL_THRESHOLD = 95;
+    private static final int PRUNE_MEMBER_AMOUNT = 10;
+    private static final Period INACTIVE_AFTER = Period.ofDays(90);
+    private static final int RECENTLY_JOINED_DAYS = 7;
+
+    private final HelpSystemHelper helper;
+    private final ModAuditLogWriter modAuditLogWriter;
+    private final Database database;
+    private final List<String> allCategories;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param config to determine all helper categories
+     * @param helper the helper to use
+     * @param modAuditLogWriter to inform mods when manual pruning becomes necessary
+     * @param database to determine whether an user is inactive
+     */
+    public AutoPruneHelperRoutine(@NotNull Config config, @NotNull HelpSystemHelper helper,
+            @NotNull ModAuditLogWriter modAuditLogWriter, @NotNull Database database) {
+        allCategories = config.getHelpSystem().getCategories();
+        this.helper = helper;
+        this.modAuditLogWriter = modAuditLogWriter;
+        this.database = database;
+    }
+
+    @Override
+    public @NotNull Schedule createSchedule() {
+        return new Schedule(ScheduleMode.FIXED_RATE, 0, 1, TimeUnit.HOURS);
+    }
+
+    @Override
+    public void runRoutine(@NotNull JDA jda) {
+        jda.getGuildCache().forEach(this::pruneForGuild);
+    }
+
+    private void pruneForGuild(@NotNull Guild guild) {
+        TextChannel overviewChannel = guild.getTextChannels()
+            .stream()
+            .filter(channel -> helper.isOverviewChannelName(channel.getName()))
+            .findAny()
+            .orElseThrow();
+        Instant when = Instant.now();
```

Comment:
```suggestion

        Instant now = Instant.now();
```
I think `now` reads better in this context, with usages:

```java

if (Duration.between(memberJoined, when).toDays() <= RECENTLY_JOINED_DAYS)
```
and

```java
Instant latestActiveMoment = when.minus(INACTIVE_AFTER);
```
Your choice, I'm just nitpicking.

Reply:
agreed